### ISSUE TITLE
-Wcast-align incorrectly warning on aligned type.

### ISSUE THREAD

**Issue by jyknight** [llvm/llvm-project#45950]
|  |  |
| --- | --- |
| Bugzilla Link | [46605](https://llvm.org/bz46605) |
| Version | trunk |
| OS | All |
| CC | @zygoloid |

## Extended Description 
This seems to be a regression from 854f5f332af4640d9425e9a94442629e4f5a3f98 ("[Sema] Teach -Wcast-align to compute an accurate alignment using the alignment information on VarDecls in more cases.")

Running `clang -Wcast-align test.cc` on the following program:

```
class Foo {
  long x;
};

struct Bar {
  alignas(Foo) char buf[sizeof(Foo)];
  Foo* getFoo() { return (Foo*)&buf; }
};
```

results in an incorrect warning, despite that "buf" is actually correctly 8-byte aligned:

test.cc:7:26: warning: cast from 'char (*)[8]' to 'Foo *' increases required alignment from 1 to 8 [-Wcast-align]
  Foo* getFoo() { return (Foo*)&buf; }
                         ^~~~~~~~~~
